app/controllers/helpers/login_utils.rb

```diff
@@ -0,0 +1,39 @@
+module LoginUtils
+
+  def current_user
+    User.find(session[:user_id])
+  end
+
+  def login(params)
+
+    user = User.find_by(username: params[:username])
+
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      flash[:message] = "Welcome, #{user.username}!"
+      redirect '/tweets'
+    else
+      flash[:message] = "Sorry. Incorrect login information."
+      redirect '/login'
+    end
+
+  end
+
+  def if_logged_in
+    if session[:user_id]
+      yield
+    else
+      flash[:message] = "Please log in."
+      redirect '/login'
+    end
+  end
+
+  def logged_in?
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```